lib/src/async_cache.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+typedef DateTime _GetNow();
+
+/// Runs asynchronous functions and caches the result for a period of time.
+///
+/// This class exists to cover the pattern of having potentially expensive code
+/// such as file I/O, network access, or isolate computation run fewer times in
+/// an application or process.
+///
+/// ## Example use with a `Future`:
+/// ```dart
+/// final _usersCache = new AsyncCache<List<String>>();
+///
+/// /// Runs the actual expensive code.
+/// Future<List<String>> _getOnlineUsers() => ...
+///
+/// /// Uses the cache if it exists, otherwise calls `_getOnlineUsers`.
+/// Future<List<String>> getOnlineUsers() => _usersCache.fetch(_getOnlineUsers);
+/// ```
+class AsyncCache<T> {
+  static DateTime _defaultNow() => new DateTime.now();
+
+  final Duration _duration;
+  final _GetNow _getNow;
+
+  List<T> _cachedStreamValues;
+  Future _fetchingStream;
+
+  Future<T> _cachedValueFuture;
+  DateTime _cachedExpiration;
+
+  /// Creates a cache that invalidates contents after [duration] has passed.
+  ///
+  /// For _testing_, a custom `now` function may be provided.
+  factory AsyncCache(Duration duration, {DateTime now()}) = AsyncCache<T>._;
```

Comment:
I guess my question is should I gate landing this with a testable API on adding a dependency on pkg/clock, which looks like it hasn't been maintained since 2014? I could also remove `now` and say manually call `invalidate` until we figure out our strategy. WDUT?